### PR TITLE
JBRes-2222 Specify Function Types Explicitly

### PR DESCRIPTION
**Pull request by SBOne-Kenobi**
## Description

The aim is to provide types of todo-replacements explicitly to avoid compilation errors.
Here are the changes that I have made:
- added kotlin type renderer that considers imports inside the file
- modified PsiBodyDDItem to store pre-rendered types
- optimized PsiTrie to store only marked elements
- other small fixes

### Automated tests

All tests are still relevant + it's hard to test type resolving normally. Most of the types are not resolved

### Manual tests

kaliningraph, but it's needed to make some additional effort, because now we have some problems with comparing the exception

## Self-check list

- [x] PR **title** and **description** are clear and aligned with a format.
- [x] I've added enough **comments** to my code, particularly in hard-to-understand areas. 
- [x] The functionality I've repaired, changed or added is covered with **automated tests**.
- [x] **Manual tests** have been provided _optionally_.
- [x] The **documentation** for the functionality I've been working on is up-to-date/provided.
- [x] The **link** to this PR is commented on in the corresponding **YT ticket**.

Hint: [x] is a marked item

### CODE REVIEW

File: project-minimization-plugin/src/test/kotlin/MinimizationServiceTest.kt
Comment (khbminus) [inline project-minimization-plugin/src/test/kotlin/MinimizationServiceTest.kt:1]: Why did you delete the test? Don't we some time in the future need it for testing? I mean if the headless will work

File: project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/model/IJDDItem.kt [hunk 1]
```diff
@@ -75,4 +82,19 @@ data class PsiWithBodyDDItem(
...
+        fun transform(accessor: KtPropertyAccessor): T
+
+        fun transform(element: PsiElement): T = when (element) {
+            is KtClassInitializer -> transform(element)
+            is KtNamedFunction -> transform(element)
+            is KtLambdaExpression -> transform(element)
+            is KtPropertyAccessor -> transform(element)
+            else -> error("Invalid PSI element type: ${element::class.simpleName}. Expected one of: KtClassInitializer, KtNamedFunction, KtLambdaExpression, KtPropertyAccessor")
+        }
+    }
```
Comment (khbminus) [inline project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/model/IJDDItem.kt:99]: Well, I quite don't understand the reasoning behind making this abstraction. I mean sure it makes sense to mark all similar classes, but why not just give up on it before it's actually needed?

I mean it's only two implementations: `PsiBodyReplacer` and `PsiBodyTypeRenderer` that share the same "affected" psi types. However, if there is another set of psis that could be transformed (I look on instance-level), then we would need some tweaking and workarounds. 

That's the same problem as when I force `PsiDDItem` to be constructable from chosen set of Psi elements

**Review by khbminus — APPROVED** [review]
The big comment is still just a comment, I'm just preparing to merge into instance-level :) LGTM

File: project-minimization-plugin/src/test/kotlin/MinimizationServiceTest.kt
Comment (SBOne-Kenobi) [reply project-minimization-plugin/src/test/kotlin/MinimizationServiceTest.kt:1]: We will rewrite it anyway, but you're right. I will create a ticket so we don't forget it.

File: project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/model/IJDDItem.kt (hunk 1 above)
Comment (SBOne-Kenobi) [reply project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/model/IJDDItem.kt:99]: I need this interface only for the body-replacement algorithm. If we change the type of elements with the bodies we are handling, we will change this interface and remember to change all parts based on the handling types. I'd rather prefer also to make the PsiWithBodyDDItem as some hierarchy, but I'm afraid that your PR will make some other changes, so I tried to keep the current version for seamless integration

Comment (SBOne-Kenobi) [reply project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/model/IJDDItem.kt:99]: But if you want I can move this interface to the other place

Comment (khbminus) [reply project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/model/IJDDItem.kt:99]: Nah, that's fine :)

**Pull request merged**